xds/internal/balancer/cdsbalancer/cdsbalancer.go

```diff
@@ -510,33 +305,357 @@ func (b *cdsBalancer) UpdateClientConnState(state balancer.ClientConnState) erro
+	for name, state := range b.watchers {
+		state.cancelWatch()
+		delete(b.watchers, name)
+	}
+}
+
 // Close cancels the CDS watch, closes the child policy and closes the
 // cdsBalancer.
 func (b *cdsBalancer) Close() {
-	b.closed.Fire()
-	<-b.done.Done()
+	b.serializer.Schedule(func(ctx context.Context) {
+		b.closeAllWatchers()
+
+		if b.childLB != nil {
+			b.childLB.Close()
+			b.childLB = nil
+		}
+		if b.cachedRoot != nil {
+			b.cachedRoot.Close()
+		}
+		if b.cachedIdentity != nil {
+			b.cachedIdentity.Close()
+		}
+		b.logger.Infof("Shutdown")
+	})
+	b.serializerCancel()
+	<-b.serializer.Done()
 }
 
 func (b *cdsBalancer) ExitIdle() {
-	b.updateCh.Put(exitIdle{})
+	b.serializer.Schedule(func(context.Context) {
+		if b.childLB == nil {
+			b.logger.Warningf("Received ExitIdle with no child policy")
+			return
+		}
+		// This implementation assumes the child balancer supports
+		// ExitIdle (but still checks for the interface's existence to
+		// avoid a panic if not).  If the child does not, no subconns
+		// will be connected.
+		if ei, ok := b.childLB.(balancer.ExitIdler); ok {
+			ei.ExitIdle()
+		}
+	})
+}
+
+// Handles a good Cluster update from the xDS client. Kicks off the discovery
+// mechanism generation process from the top-level cluster and if the cluster
+// graph is resolved, generated child policy config and pushes it down.
+//
+// Only executed in the context of a serializer callback.
+func (b *cdsBalancer) onClusterUpdate(name string, update xdsresource.ClusterUpdate) {
+	state := b.watchers[name]
+	if state == nil {
+		// We are currently not watching this cluster anymore. Return early.
+		return
+	}
+
+	b.logger.Infof("Received Cluster resource: %s", pretty.ToJSON(update))
+
+	// Update the watchers map with the update for the cluster.
+	state.lastUpdate = &update
+
+	// For an aggregate cluster, always use the security configuration on the
+	// root cluster.
+	if name == b.lbCfg.ClusterName {
+		// Process the security config from the received update before building the
+		// child policy or forwarding the update to it. We do this because the child
+		// policy may try to create a new subConn inline. Processing the security
+		// configuration here and setting up the handshakeInfo will make sure that
+		// such attempts are handled properly.
+		if err := b.handleSecurityConfig(update.SecurityCfg); err != nil {
+			// If the security config is invalid, for example, if the provider
+			// instance is not found in the bootstrap config, we need to put the
+			// channel in transient failure.
+			b.logger.Warningf("Received Cluster resource contains invalid security config: %v", err)
+			b.onClusterError(name, err)
+			return
+		}
+	}
+
+	clustersSeen := make(map[string]bool)
+	dms, ok, err := b.generateDMsForCluster(b.lbCfg.ClusterName, 0, nil, clustersSeen)
+	if err != nil {
+		b.logger.Warningf("Failed to generate discovery mechanisms: %v", err)
+		b.onClusterError(b.lbCfg.ClusterName, err)
```

Comment:
Line 450 and 451 send the same error string down as the logging string. I think the prefix "Failed to generate discovery mechanisms" adds valuable information here.

Reply:
Realized that `onClusterError` also logs the error. So, got rid of the log statement here, and in other places where I was logging and passing the error to `onClusterError`.